clang/lib/Sema/SemaExpr.cpp

```diff
@@ -262,12 +262,32 @@ bool Sema::DiagnoseUseOfDecl(NamedDecl *D, ArrayRef<SourceLocation> Locs,
       // there are some exceptions from this rule - functions that start
       // with '__spirv_', '__sycl_', '__assert_fail', etc.
       const IdentifierInfo *Id = FDecl->getIdentifier();
+      auto isMsvcMathFn = [=](StringRef Name) {
```

Comment:
We need to rethink this diagnostic. The design of this diagnostic is strange in my opinion. "Do not allow SYCL device functions in ESIMD context excepts for func1,func2, etc". I don't think something like this will be upstreamable.

Also, how likely is it that this list will just keep growing? Can we just avoid the diagnostic for builtins instead? 

@premanandrao @AaronBallman @tahonermann can you please weigh in here?

Reply:
Good points, @elizabethandrews.  I agree.   It seems like a very arbitrary list to me.

Reply:
Please correct me, but the upstreaming plans did not include ESIMD. 

Also, I agree that we need to rather use some sort of attributes telling that the function is usable/implemented for ESIMD context. It remains in our plans, but has lower priority than several feature requests that we already have and are working on.

Reply:
I agree with above points. The list is not expected to grow significantly, device library does not seem to change that often. This PR, however, is important to cover one critical request from our customers.

Reply:
I'm actually not sure about ESIMD upstreaming. @premanandrao might know. If we're not upstreaming this, I guess I am less concerned but this still makes me uneasy (we may not need to listen to my uneasiness based on how @premanandrao / @smanna12  / @AaronBallman feels).

>> Also, I agree that we need to rather use some sort of attributes telling that the function is usable/implemented for ESIMD context.

Maybe I am missing something but I'm not sure how attributes will help if functions are declared/defined in other libraries/headers where we cannot add the attributes.

Reply:
> I'm actually not sure about ESIMD upstreaming. @premanandrao might know. If we're not upstreaming this, I guess I am less concerned but this still makes me uneasy (we may not need to listen to my uneasiness based on how @premanandrao / @smanna12 / @AaronBallman feels).

I've seen no community RFC for esimd and it wasn't included in the scope of the SYCL upstreaming RFCs, so my assumption is it's not on the table for upstreaming currently.

In terms of uneasiness, I think the list is somewhat hacky but the underlying concept is actually what concerns me. The list is not a complete list of floating-point primitives (https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/floating-point-primitives?view=msvc-170) and it includes entries that are not documented by Microsoft. Further, Microsoft recommends against using these interfaces at all:

> Microsoft-specific primitive functions that are used to implement some standard C runtime library (CRT) floating-point functions. They're documented here for completeness, but aren't recommended for use. Some of these functions are noted as unused, because they're known to have issues in precision, exception handling, and conformance to IEEE-754 behavior. They exist in the library only for backward compatibility. For correct behavior, portability, and adherence to standards, prefer the standard floating-point functions over these functions.

So I have no idea what gets onto this list and why which makes it a minor maintenance concern. That said, if the esimd folks think this is The List they need, the implementation is reasonable enough.

Reply:
The list only includes primitive functions that are included into device library.
These functions are not for the developer to be used directly, but through the Microsoft STL functions.
The current ESIMD approach is to highlight the unsupported functions, and pass through only the supported ones.

Reply:
The part that leaves me uncomfortable is that we're claiming to support functions which are private interfaces we don't know anything about. e.g., `_FDtest`. If the list was just the stuff that's documented, then it makes more sense to me.

Reply:
I don't know the device library design so I can't speak to whether that design is reasonable or not (from the outside looking in, it seems unreasonable to make assumptions about other people's undocumented APIs and claim any kind of support for them). I think the idea itself is perhaps a bit of a hack, but I don't have technical concerns with these specific changes as far as they go.

If the SYCL and ESIMD people are comfortable with this design, I won't block it.

Reply:
I am not comfortable with it, but I can't think of any good alternatives, so I will go ahead and approve it, if @elizabethandrews is okay too.

Reply:
I echo everyone's feelings about being (uncomfortably) ok because I too do not know device library design and have no better ideas.

Reply:
>>I echo everyone's feelings about being (uncomfortably) ok because I too do not know device library design and have no better ideas.

I have same opinion.